Demo and fix broken pagination

sonthonaxrk opened the pull request:
Originally discussed in #7888

This demonstrates how the pagination is broken if someone tries to order by `non_unique_field,id`.  The only case where it does work is when the offset_cutoff is the same as the page_size. 

```
$ pytest tests/test_cursor_pagination.py -v

tests/test_cursor_pagination.py::test_filtered_items_are_paginated[page_size_divisor_of_offset] FAILED                                             [ 20%]
tests/test_cursor_pagination.py::test_filtered_items_are_paginated[page_size_multiple_of_offset] FAILED                                            [ 40%]
tests/test_cursor_pagination.py::test_filtered_items_are_paginated[page_size_uneven_divisor_of_offset] FAILED                                      [ 60%]
tests/test_cursor_pagination.py::test_filtered_items_are_paginated[page_size_uneven_multiple_of_offset] FAILED                                     [ 80%]
tests/test_cursor_pagination.py::test_filtered_items_are_paginated[page_size_same_as_offset] PASSED                                                [100%]



```

sonthonaxrk commented on tests/test_pagination.py:
```diff
@@ -700,52 +781,6 @@ def test_cursor_pagination(self):
-        assert next is None
-        assert previous_url is not None
-        assert next_url is None
-
```
This test was a bit too complicated to test with the `MockQueryset`.

sonthonaxrk commented on tests/test_pagination.py:
```diff
@@ -884,7 +919,7 @@ def test_cursor_pagination_with_page_size_negative(self):
         assert previous == [4, 4, 5, 6, 7]
         assert current == [7, 7, 7, 7, 7]
-        assert next == [8, 9, 9, 9, 9]  # Paging artifact
+        #assert next == [8, 9, 9, 9, 9]  # Paging artifact
```
I'm pretty sure these 'paging artefacts' have been fixed by this.

sonthonaxrk commented on tests/test_pagination.py:
```diff
@@ -13,6 +21,79 @@
+
+            return _statements
+
+        return lambda item: operator_map[q_object.connector](l(item) for l in _parse(q_object))
```
I'm not a massive fan of 'MockQuerysets' because we're really reimplementing stuff that's handled by the database.  But I tried to keep the pattern of tests as they are.

tomchristie commented:
Okay, so this PR falls a long way into our "no" territory.

You've spent time and care on this, so I'll try to help explain some of why our outlook on this as maintainers differs to yours.

> This demonstrates how the pagination is broken if someone tries to order by non_unique_field,id.

So, the pull request doesn't demonstrate much to me because it's a ~430 line pull request that'd require a huge amount of time to fully review, for a use case that doesn't really meet the documented constraints of the CursorPagination.

To give some contrast, here's a one-liner pull request, that seemed pretty reasonable (use deque instead of list, for performance reasons) which we accepted, and then reverted once it turned out that it actually broke some existing usages for unexpected reasons.

When we have to spend a large amount of time in order to assess if a 1 line PR meets our expected requirements for code changes in REST framework, I'd hope its clear that large scoped pull requests like this just aren't realistic, particularly when the motivation and understanding of the trade-offs it introduces haven't first been adequately addressed in discussion.

If you're invested in putting time into this, then the best route is as a third party package, which we would then link to from the main docs... https://www.django-rest-framework.org/api-guide/pagination/#third-party-packages

sonthonaxrk commented:
@tomchristie  I tried to highlight was exactly was wrong with by posting a test, and I felt that was ignored and simply told I was 'using it wrong'.  I get it's a little bit subtle and doesn't happen in all cases; and I also understand that you probably have 100s of inexperienced developers vying for your time who have no idea what they're doing. 

I know this stuff is taxing to read.  As per the project guidelines, I submitted a pull request, first with the failing test, then another commit with the fix.  I even linked the individual commits (for the problem, and the fix).  But I don't think anyone even took the time to examine how it's broken (because it's hard work) and instead knee-jerked with, it's not a problem and that even if it was, I should read the documentation.

Look, I totally get the tedium of managing a package that's used by thousands, and how APIs are fragile.  I've managed packages too.  However, it would have been nice to at least have had the reaction "yes this is bug, but we can't fix it because of x".  Instead I've just been told that I'm using it wrong.  With respect, I'm actually working on a project that you previously consulted on, and this is an edge case that you were personally caught out by (and partly why I felt that I should contribute back).

---

The only reason that I added more commits was that my change broke the old tests (because the tests seem to both run as mocks and with real objects).  Because I was told that I was 'using it wrong' and not reading the documentation, I started to doubt myself, and spent a few hours verifying that my change was in fact okay.  I haven't meaningfully changed the tests.  In fact I think I fixed a few subtle bugs that are labeled 'paging artefact'.

tomchristie commented:
Ok noted. I'm too busy to have a second look over this right now, but can take a re-review next week sometime.

I guess the helpful thing from my POV would be starting with an absolutely minimal PR, that demonstrates a breakage with as *absolutely small a change footprint as possible*. (Although I realise that's asking for more work on your side)

sonthonaxrk commented:
Thanks @tomchristie, I really appreciate that.  I sorry for being curt, I was just a bit frustrated there.  I also understand that contributing to open source takes time, and that problems often languish for months before they can be merged

In regards to the breakage, I tried to keep it small (while being comprehensive).  There's only one test there, I just I would be able to write clearer code if the problem was encapsulated in a single file.

https://github.com/sonthonaxrk/django-rest-framework/blob/29d8796b1d96cbe77ecd81663ee7afbace0229e0/tests/test_cursor_pagination.py
